cpp/src/parquet/encoding_test.cc

```diff
@@ -862,5 +862,215 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
+  ASSERT_EQ(0, encoded_buffer->size());
+}
+
+// Check that the decoder can handle empty input.
+TEST(ByteStreamSplitEncodeDecode, DecodeZeroLenInput) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(0, NULL, 0);
+  ASSERT_EQ(0U, decoder->Decode(NULL, 0));
+}
+
+TEST(ByteStreamSplitEncodeDecode, DecodeOneLenInput) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U};
+  decoder->SetData(1, data, 4);
+
+  float value = .0f;
+  const int num_decoded = decoder->Decode(&value, 1);
+  ASSERT_EQ(1, num_decoded);
+
+  const uint32_t value_as_uint = *reinterpret_cast<const uint32_t*>(&value);
+  ASSERT_EQ(0x1337C0DEU, value_as_uint);
+}
+
+// Check that requesting to decode more elements than is available in the storage
+// of the decoder works correctly.
+TEST(ByteStreamSplitEncodeDecode, DecodeLargerPortion) {
+  std::unique_ptr<TypedDecoder<DoubleType>> decoder =
+      MakeTypedDecoder<DoubleType>(Encoding::BYTE_STREAM_SPLIT);
+  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U, 0x11U, 0x22U, 0x33U, 0x44U,
+                          0xAAU, 0xBBU, 0xCCU, 0xDDU, 0x55U, 0x66U, 0x77U, 0x88U};
+  decoder->SetData(sizeof(data) / sizeof(double), data, sizeof(data));
+
+  double values[2] = {.0, .0};
+  const int num_decoded = decoder->Decode(values, 10000);
+  ASSERT_EQ(2, num_decoded);
+
+  uint64_t value_as_uint = *reinterpret_cast<const uint64_t*>(&values[0]);
+  ASSERT_EQ(static_cast<uint64_t>(0x7755CCAA331137DEULL), value_as_uint);
+
+  value_as_uint = *reinterpret_cast<const uint64_t*>(&values[1]);
+  ASSERT_EQ(static_cast<uint64_t>(0x8866DDBB442213C0ULL), value_as_uint);
+}
+
+// Check that the decoder can decode the input in smaller steps.
+TEST(ByteStreamSplitEncodeDecode, DecodeMultipleTimes) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  const int num_values = 100;
+  std::vector<uint8_t> data(num_values * 4);
+  for (size_t i = 0; i < data.size(); ++i) {
+    data[i] = static_cast<uint8_t>(i & 0xFFU);
+  }
+  decoder->SetData(num_values, data.data(), num_values * 4);
+
+  const int step = 25;
+  std::vector<float> decoded_data(step);
+  for (int i = 0; i < num_values; i += step) {
+    int num_decoded = decoder->Decode(decoded_data.data(), step);
+    ASSERT_EQ(step, num_decoded);
+    for (int j = 0; j < step; ++j) {
+      const uint32_t assembled_value =
+          static_cast<uint32_t>(data[i + j]) |
+          (static_cast<uint32_t>(data[(i + j) + num_values]) << 8U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 2]) << 16U) |
+          (static_cast<uint32_t>(data[(i + j) + num_values * 3]) << 24U);
+      const float assembled_value_as_float =
+          *reinterpret_cast<const float*>(&assembled_value);
```

Comment:
Is there value in this manual reassembly versus a static or randomly generated array?

Reply:
If we test the round-trip Decode(Encode(A)) == A then we only test that the decoder does the inverse of the encoder. Since we have a spec we have to follow, it's good to check that everything is precise.
If we get a failing test, we might not know whether it's in the decode or encode path. Testing only the encoder/decoder can help root cause the issue.